src/lib/parsers/sol-json.c

```diff
@@ -789,6 +790,492 @@ sol_json_serialize_boolean(struct sol_buffer *buffer, bool val)
+                .start = SOL_STR_SLICE_LITERAL("{"),
+                .end = SOL_STR_SLICE_LITERAL("}"),
+            },
+            .key = {
+                .end = SOL_STR_SLICE_LITERAL(":"),
+            },
+            .separator = SOL_STR_SLICE_LITERAL(","),
+            .show_key = true,
+            .detailed = detailed_structures,
+        },
+        .array = {
+            .container = {
+                .start = SOL_STR_SLICE_LITERAL("["),
+                .end = SOL_STR_SLICE_LITERAL("]"),
+            },
+            .separator = SOL_STR_SLICE_LITERAL(","),
+            .show_index = false,
+        },
+    };
+
+    SOL_NULL_CHECK(buffer, -EINVAL);
+    SOL_NULL_CHECK(desc, -EINVAL);
+    SOL_NULL_CHECK(memory, -EINVAL);
+
+    return sol_memdesc_serialize(desc, memory, buffer, &json_serialize_options, NULL);
+}
+
+static inline const struct sol_memdesc_structure_member *
+find_memdesc_for_key(const struct sol_memdesc *desc, const struct sol_json_token *token)
+{
+    const struct sol_memdesc_structure_member *found;
+    struct sol_buffer buf;
+    struct sol_str_slice key;
+
+    if (sol_json_token_get_unescaped_string(token, &buf) < 0)
+        return NULL;
+
+    key = sol_buffer_get_slice(&buf);
+    found = sol_memdesc_find_structure_member(desc, key);
+    sol_buffer_fini(&buf);
+    return found;
+}
+
+SOL_API int
+sol_json_load_memdesc(const struct sol_json_token *token, const struct sol_memdesc *desc, void *memory)
+{
+    int r;
+
+    SOL_NULL_CHECK(token, -EINVAL);
+    SOL_NULL_CHECK(desc, -EINVAL);
+    SOL_NULL_CHECK(memory, -EINVAL);
+
+    switch (desc->type) {
+    case SOL_MEMDESC_TYPE_UNKNOWN:
+        return -EINVAL;
+    case SOL_MEMDESC_TYPE_UINT8: {
+        uint8_t *m = memory;
+        uint64_t v;
+
+        r = sol_json_token_get_uint64(token, &v);
+        if (r < 0)
+            return r;
+
+        if (v > UINT8_MAX)
+            return -ERANGE;
+
+        *m = v;
+        return 0;
+    }
+    case SOL_MEMDESC_TYPE_UINT16: {
+        uint16_t *m = memory;
+        uint64_t v;
+
+        r = sol_json_token_get_uint64(token, &v);
+        if (r < 0)
+            return r;
+
+        if (v > UINT16_MAX)
+            return -ERANGE;
+
+        *m = v;
+        return 0;
+    }
+    case SOL_MEMDESC_TYPE_UINT32:
+        return sol_json_token_get_uint32(token, memory);
+    case SOL_MEMDESC_TYPE_UINT64:
+        return sol_json_token_get_uint64(token, memory);
+    case SOL_MEMDESC_TYPE_ULONG: {
+        unsigned long *m = memory;
+        uint64_t v;
+
+        r = sol_json_token_get_uint64(token, &v);
+        if (r < 0)
+            return r;
+
+        if (v > ULONG_MAX)
+            return -ERANGE;
+
+        *m = v;
+        return 0;
+    }
+    case SOL_MEMDESC_TYPE_SIZE: {
+        size_t *m = memory;
+        uint64_t v;
+
+        r = sol_json_token_get_uint64(token, &v);
+        if (r < 0)
+            return r;
+
+        if (v > SIZE_MAX)
+            return -ERANGE;
```

Comment:
You are comparing unsigned `v` with signed `SIZE_MAX`. We are appending `Wsign-compare` if available to `CFLAGS`. Compilers will complain with this comparation. There are many other places doing it.